### ISSUE TITLE
Path in default .addins file for ConsoleRunner package may not exist

### ISSUE THREAD

**Issue by nathan-alden** [nunit/nunit#1551]
I'm attempting to use the files packaged in the `NUnit.ConsoleRunner` 3.2.1 NuGet package on [nuget.org](https://www.nuget.org/packages/NUnit.ConsoleRunner). Unfortunately, my TeamCity builds are failing because `nunit3-console.exe` is throwing a `NullReferenceException`:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Engine.Internal.DirectoryFinder.ExpandOneStep(IList`1 dirList, String pattern)
   at NUnit.Engine.Internal.DirectoryFinder.GetDirectories(DirectoryInfo baseDir, String pattern)
   at NUnit.Engine.Services.ExtensionService.ProcessAddinsFile(DirectoryInfo baseDir, String fileName)
   at NUnit.Engine.Services.ExtensionService.FindExtensionsInDirectory(DirectoryInfo startDir)
   at NUnit.Engine.Services.ExtensionService.StartService()
   at NUnit.Engine.Services.ServiceManager.StartServices()
   at NUnit.Engine.TestEngine.NUnit.Engine.ITestEngine.get_Services()
   at NUnit.ConsoleRunner.ConsoleRunner..ctor(ITestEngine engine, ConsoleOptions options, ExtendedTextWriter writer)
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```

Here's the command line TeamCity is executing:

```
C:\NUnit\nunit3-console.exe C:\TeamCity\buildAgent\temp\buildTmp\fHz0v8K0KeWMqFUeAGIjXbAG22pExfA3.nunit --work=C:\TeamCity\buildAgent\work\b967e7acecd222e2 --noresult --noheader
```

Just to be sure this wasn't a problem with TeamCity itself, I ran `nunit3-console.exe` myself on the build agent and passed in the test assembly directly, rather than using an NUnit project file like TeamCity:

```
C:\NUnit\nunit3-console.exe C:\TeamCity\buildAgent\work\b967e7acecd222e2\src\UnitTests\bin\Release\UnitTests.dll
```

No dice--the same exception was thrown.

Next, I downloaded the latest source from https://github.com/nunit/nunit.git, compiled the solution in Debug configuration, and copied the necessary files plus PDBs to the build agent:

![image](https://cloud.githubusercontent.com/assets/1189759/15881016/375fe918-2cf8-11e6-98c6-52e9c1af9e14.png)

My hope was to get a line number in the `ExpandOneStep` method so I could investigate further. Unfortunately (fortunately?), this time the build worked! I don't have any tests yet defined in `UnitTests`, so the `Invalid` error is expected:

```
C:\NUnit\nunit3-console.exe C:\TeamCity\buildAgent\work\b967e7acecd222e2\src\UnitTests\bin\Release\UnitTests.dll
NUnit Console Runner 3.3.0 (Debug)
Copyright (C) 2016 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 6.3.9600.0
  CLR Version: 4.0.30319.42000

Test Files
    C:\TeamCity\buildAgent\work\b967e7acecd222e2\src\UnitTests\bin\Release\UnitTests.dll


Errors and Failures

1) Invalid : C:\TeamCity\buildAgent\work\b967e7acecd222e2\src\UnitTests\bin\Release\UnitTests.dll
No suitable tests found in 'C:\TeamCity\buildAgent\work\b967e7acecd222e2\src\UnitTests\bin\Release\UnitTests.dll'.
Either assembly contains no tests or proper test driver has not been found.

Test Run Summary
  Overall result: Failed
  Test Count: 0, Passed: 0, Failed: 0, Inconclusive: 0, Skipped: 0
  Start time: 2016-06-08 02:28:21Z
    End time: 2016-06-08 02:28:21Z
    Duration: 0.004 seconds

Results (nunit3) saved as TestResult.xml
```

I then tried to run the original command that TeamCity executed. Again unfortunately, this time I avoided the `NullReferenceException` but got a different error, instead:

```
1) Invalid : C:\TeamCity\buildAgent\temp\buildTmp\fHz0v8K0KeWMqFUeAGIjXbAG22pExfA3.nunit
File type is not supported
```

I am not sure what the differences are between what's in the `NUnit.ConsoleRunner` 3.2.1 package and what's in Git, but you may want to take a look at this. It seems that no matter what I try I cannot get this TeamCity integration working. It does not appear, however, that the problem lies with TeamCity.

**Comment by CharliePoole** [comment]
The stack trace suggests that there may be an invalid `.addins` file somewhere. Your installation of the NUnit.ConsoleRunner package should include nunit.engine.addins with content

```
../../NUnit.Extension.*/tools/
```

The directories located by this path may contain other `.addins` files, recursively.

**Comment by nathan-alden** [comment]
Yes, the `nunit.nuget.addins` file that ships inside `NUnit.ConsoleRunner` does have that line. However, the code that processes those addins does not seem resilient against directories that don't exist. I have the console runner located in `C:\NUnit` so there is no `../../` folder. It almost seems like `NUnit.ConsoleRunner` is not designed to be used independently from some other installation of NUnit.

My assumption was that if one simply wanted to use the console runner for a CI environment, one should simply grab `NUnit.ConsoleRunner` from NuGet. Is this not recommended? Regardless of the answer, it does seem strange to have this fixed path requirement in the `nunit.nuget.addins` file.

**Comment by nathan-alden** [comment]
I installed the NUnit 3.2.1 installation on my TeamCity server and re-pointed TeamCity to that new location. Seems like the build is working well now. I'll avoid using the NuGet packages directly in the future.

**Comment by ChrisMaddock** [comment]
Is there not a couple of separate bugs here?

1. DirectoryFinder can thrown on an invalid path, 
2 The path stated in nunit.nuget.addins may not exist if the package is installed only one directory deep

The latter will throw an `NUnitEngineException` I believe. 

@CharliePoole - I've reopened and retitled, please feel free to close again if I've missed something, or you think these fall under wontfix. :-)

**Comment by CharliePoole** [comment]
@ChrisMaddock This shouldn't be closed. I was waiting for more info from the poster but my last comment does not seem to have registered with GitHub. 

@nathan-alden The line in the addins file is based on the notion that nuget will install the binaries in some directory like C:\NUnit\packages\NUnit.ConsoleRunner. So if it's somehow not doing that we need to rethink things. How did the binaries get to there location on your server?

**Comment by CharliePoole** [comment]
@ChrisMaddock Both of your points are correct. We need to check in the code. Additionally, we may need to revise our thinking about how to find extensions installed as a package. The current wildcard directory search was added as a first cut and works most of the time but we may need something more.

**Comment by nathan-alden** [comment]
Sorry for closing the issue, Charlie. I wasn't aware that work might come out of it.

**Comment by CharliePoole** [comment]
@nathan-alden It's only about a week till the 3.4 release. If you can answer my two-comments-back question we may be able to do something with this, but otherwise it will wait another three months.

**Comment by nathan-alden** [comment]
I just extracted the `.nupkg` with 7-Zip, then dropped the binaries onto my TeamCity server in the `C:\NUnit` directory.

**Comment by CharliePoole** [comment]
OK, so that's not the normal nuget install, but I wonder if other people do that. We do have a zip package and it doesn't have the `.addins` file that messed you up.

My conclusion is that we don't have to make the search for extensions work in your situation, just that it shouldn't crash. Would you agree?

**Comment by nathan-alden** [comment]
Just to be sure there weren't any differences, I ran `NuGet.exe install NUnit.ConsoleRunner`. It's my understanding that all NuGet does with the `install` command is extract the .nupkg and maybe run a PowerShell script. In the case of `NUnit.ConsoleRunner`, I ended up with the same `tools` folder and a `nunit.nuget.addins` that had the `..\..\` path.

Yeah, I think that if the directory-navigating class simply handled exceptions, I probably never would've even noticed that the `.addins` file was causing a problem.
